### DISCUSSION TITLE
response_model_skip_defaults doesn't appear to be working?

### DISCUSSION THREAD

**Discussion by wozniakty** [fastapi/fastapi#8206] [Questions] [answered]
**Describe the bug**
It's pretty straightforward, I've used the `response_model_skip_defaults` flag but it's returning the full model anyways.  I'll try to post code examples but it's really barebones.

**To Reproduce**
1. Create a file with '...'
2. Add a path operation function with '....'
3. Open the browser and call it with a payload of '....'

**Code**
Across a few different files...
```
App = FastAPI(debug=settings.debug)
App.add_middleware(RequestContextMiddleware)
App.add_middleware(ApiTokenMiddleware)
App.add_middleware(SentryMiddleware)
App.include_router(profile_router,
    prefix='/org/{org_id:int}/groups/{group_id:int}')

...

@validate_token(scope='usermanagement')
@profile_router.get('/profile', response_model=Profile, response_model_skip_defaults=True)
async def get_profile(org_id: int, group_id: int):
    return Profile(description="These are words")

...

class ProfileContact(BaseModel):
    email: EmailStr = None
    phone: str = None
    website: str = None

class ProfileAddress(BaseModel):
    address: str = None
    city: str = None
    state: str = None
    zip: str = None
    country: str = None

class Profile(BaseModel):
    first_name: str = None
    last_name: str = None
    title: str = None
    company: str = None
    licenses: str = None
    headline: str = None
    description: str = None
    disclaimer: str = None
    contact: ProfileContact = None
    address: ProfileAddress = None
```

some shell stuff that works just fine:
```
In [1]: from profile_management.service.profiles.records import Profile
In [2]: x = Profile(description="Test stuff")

In [3]: x.dict()
Out[3]: 
{'first_name': None,
 'last_name': None,
 'title': None,
 'company': None,
 'licenses': None,
 'headline': None,
 'description': 'Test stuff',
 'disclaimer': None,
 'contact': None,
 'address': None}

In [4]: x.dict(skip_defaults=True)
Out[4]: {'description': 'Test stuff'}

In [5]: from fastapi.encoders import jsonable_encoder

In [6]: jsonable_encoder(x, skip_defaults=True)
Out[6]: {'description': 'Test stuff'}
```

Not sure why the same behavior isn't working at the router level

**Environment:**
 - OS: Windows
 - FastAPI Version: 0.33.0
 - Python version: 3.7.3

**Comment by wozniakty** [comment]
I feel like I'm missing something super obvious.  I tried digging through the code and it looks like the param is passed around appropriately, but 🤷‍♂

**Comment by dmontagu** [comment]
Does it work if you replace the current return with `return dict(description="These are words")`? Or the same problem? (Wondering if it might be converting to dict and then back somewhere, ignoring the skip_defaults parameter, which would then make them non-default)

**Comment by wozniakty** [comment]
It works correctly with the dict, good call.  So it seems like there's a problem specifically when returning the full model type itself :|

**Comment by wozniakty** [comment]
Could be something resulting from this maybe? https://github.com/tiangolo/fastapi/blob/6ca3ce80e431c94cab4db257095e6439e5245aca/fastapi/utils.py#L53

**Comment by dmontagu** [comment]
@wozniakty I can’t confirm at just this moment, but I believe I am using this pattern in my own code and I don’t think it has the issue. I’ll check and report back to confirm whether it’s a bug or maybe an issue somewhere else.

**Comment by dmontagu** [comment]
@wozniakty I looked into it and I think this is buggy behavior. Here's a pretty minimal failing test case:

```python
from typing import Optional

from fastapi import FastAPI
from pydantic import BaseModel
from starlette.testclient import TestClient


def test_return_defaults():
    app = FastAPI()

    class Model(BaseModel):
        x: Optional[int]

    @app.get("/", response_model=Model, response_model_skip_defaults=True)
    def get_x() -> Model:
        return Model()  # adding `.dict(skip_defaults=True)` causes the test to pass

    assert TestClient(app).get("/").content == b"{}"
    # Fail: b'{"x":null}' != b'{}'
```
I suspect you are right about the source of the issue, but it may not be too hard to fix; I'm looking into it now.

**Comment by dmontagu** [comment]
Okay, I've isolated the line in FastAPI where this is happening:

https://github.com/tiangolo/fastapi/blob/6ca3ce80e431c94cab4db257095e6439e5245aca/fastapi/routing.py#L48

It looks like calling `field.validate` on a pydantic model causes the fields to become set. I'm not sure whether this is a bug in pydantic or expected. But there is two line modification to FastAPI that makes it work, which I'm submitting now as a pull request. (If it's better to fix in pydantic though, we can wait for that.)

**Comment by wozniakty** [comment]
Nice! Awesome find.  I should have been able to pull this down and fix this myself eventually, but I appreciate the help @dmontagu :+1:

**Comment by tiangolo** [comment, marked as answer]
Thanks for reporting it @wozniakty !

> I should have been able to pull this down and fix this myself eventually [...]

No worries, I'm sure that was indeed very hard to debug, it was deep inside of Pydantic's implementation, which is not simple. But @dmontagu is quite a Pydantic and FastAPI wizard by now. :sparkles: :rocket: 

You can find the fix in FastAPI `0.36.0`. :tada: 

---

Thanks @dmontagu for debugging it and finding the fix! I'm sure that was quite hard to debug, great job! :clap: :cake: 

---

I'll close this now, but feel free to add more comments or issues.

**Comment by outofnames** [comment]
Hi!
I have the following code:
```
@router.post("/token", response_model=Refresh_Access_Token, response_model_skip_defaults=True)
```
Which produces this error:
```
TypeError: post() got an unexpected keyword argument 'response_model_skip_defaults'
```
Am I missing something?

**Comment by dmontagu** [comment]
@outofnames what version of fastapi are you using? You may need to upgrade?

**Comment by outofnames** [comment]
I'll try to download the official docker image once again and report back!

**Comment by dmontagu** [comment]
The official docker image may not be up to date -- this was fixed pretty recently. If you run into trouble, tell us the value of `fastapi.__version__` -- I think this was fixed in 0.35.0.

**Comment by tiangolo** [comment]
Thanks for the help @dmontagu ! 🚀

@outofnames it would also be a good idea to install FastAPI by hand in your Dockerfile to make sure you have an updated/fixed version.

Otherwise a future version with a breaking change could break your app.

**Comment by duz-sg** [comment]
I have encountered this issue, but the trigger condition is a bit different.
In my case, I have below code
```
class ItemBase(BaseModel):
    class Config:
        orm_mode = True

class ItemCheck(ItemBase):
    customer: str = None
    category: str = None
    note: str = None

@router.get("/", response_model=ItemCheck, response_model_skip_defaults=True, status_code=200)
    check_item = crud.item.get_by_customer_ref(db_session=db, customer=customer)
    return check_item
```
The difference is, I'm returning pydantic model which is converted from a db model, using the orm_mode. 
The response I received contains all the items defined in the pydantic model, even though they are all null.
Does this considered a bug?

**Comment by dmontagu** [comment]
@duz-sg This is not a bug per-se; `orm_mode` will load all fields, and doesn't have a way to tell whether some should be left as defaults (if you have an idea about how the logic should be changed, that would be a pydantic feature request).

If you just want to remove nulls, I think the easiest way now would be to call `fastapi.encoders.jsonable_encoder` on the appropriate object(s) with `include_none=False`; basically, manually generating the pre-encoded object (rather than relying on fastapi to do it for you).

**Comment by duz-sg** [comment]
@dmontagu Thanks for your prompt reply, will look into the `jsonable_encoder` solution.
